torch/nn/modules/module.py

```diff
@@ -1509,78 +1527,105 @@ def _call_impl(self, *args, **kwargs):
+                                args, kwargs = args_kwargs_result
+                            else:
+                                raise RuntimeError(
+                                    "forward pre-hook must return None or a tuple "
+                                    f"of (new_args, new_kwargs), but got {args_kwargs_result}."
+                                )
+                    else:
+                        args_result = hook(self, args)
+                        if args_result is not None:
+                            if not isinstance(args_result, tuple):
+                                args_result = (args_result,)
+                            args = args_result
+
+            bw_hook = None
+            if full_backward_hooks or backward_pre_hooks:
+                bw_hook = hooks.BackwardHook(self, full_backward_hooks, backward_pre_hooks)
+                args = bw_hook.setup_input_hook(args)
+
+            result = forward_call(*args, **kwargs)
+            if _global_forward_hooks or self._forward_hooks:
+                for hook_id, hook in (
+                    *_global_forward_hooks.items(),
+                    *self._forward_hooks.items(),
+                ):
+                    if hook_id in self._forward_hooks_with_kwargs:
+                        hook_result = hook(self, args, kwargs, result)
+                    else:
+                        hook_result = hook(self, args, result)
+
+                    # mark that forced hook has been run
+                    if hook_id in self._forward_hooks_forced:
+                        self._forward_hooks_forced[hook_id] = True
+                    elif hook_id in _global_forward_hooks_forced:
+                        _global_forward_hooks_forced[hook_id] = True
+
+                    if hook_result is not None:
+                        result = hook_result
+
+            if bw_hook:
+                if not isinstance(result, (torch.Tensor, tuple)):
+                    warnings.warn("For backward hooks to be called,"
+                                  " module output should be a Tensor or a tuple of Tensors"
+                                  f" but received {type(result)}")
+                result = bw_hook.setup_output_hook(result)
+
+            # Handle the non-full backward hooks
+            if non_full_backward_hooks:
+                var = result
+                while not isinstance(var, torch.Tensor):
+                    if isinstance(var, dict):
+                        var = next((v for v in var.values() if isinstance(v, torch.Tensor)))
+                    else:
+                        var = var[0]
+                grad_fn = var.grad_fn
+                if grad_fn is not None:
+                    for hook in non_full_backward_hooks:
+                        grad_fn.register_hook(_WrappedHook(hook, self))
+                    self._maybe_warn_non_full_backward_hook(args, result, grad_fn)
+
+            return result
+
+        finally:
```

Comment:
I don't think this is achieving what you want. You'll trigger everything twice when no error happened?
Also if one of the hook raises an error, the next one will not be triggered?

Reply:
Maybe I am missing something here but I don't think I would trigger them twice because I have a dictionary that tracks whether a hook has fired [`_forward_hooks_forced`](https://github.com/pytorch/pytorch/blob/d1037d011854cb4ebd699ca764d637c5ac866b6d/torch/nn/modules/module.py#L1580-L1585) (and correspondingly for global forward hooks), and [I only run the hook in the `finally` block if it is marked as not having fired ](https://github.com/pytorch/pytorch/blob/d1037d011854cb4ebd699ca764d637c5ac866b6d/torch/nn/modules/module.py#L1621)

You are right that if one of the hooks in the finally block raises an error the next one will not be triggered. I will fix that!

Reply:
You are not allowed to use `self._forward_hooks_forced` for this, that will go very bad if you have concurrent forwards happening!

Reply:
Ah, so I should track the ids hat have fired within the forward call. That makes sense!